### PR TITLE
[db] Share db for two contract indexer

### PR DESCRIPTION
**Pull request by envestcc**
# Description

We want to combine db files of two staking contract indexer, which helps to make easy to maintain.

I introduce the new KVStoreShards to make different shard(with different bucket prefix of a same kvstore) of the kvstore. Therefore, the two indexer can use different shard of same kvstore.

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [x] Code refactor or improvement

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: chainservice/builder.go [hunk 1]
```diff
@@ -324,52 +324,50 @@ func (builder *Builder) buildContractStakingIndexer(forTest bool) error {
...
-
-func (builder *Builder) buildContractStakingIndexerV2(forTest bool) error {
-	if !builder.cfg.Chain.EnableStakingProtocol {
-		return nil
-	}
-	if builder.cs.contractStakingIndexerV2 != nil {
-		return nil
+	kvss := db.NewKVStoreShards(db.NewBoltDB(dbConfig))
+	// build contract staking indexer
+	if builder.cs.contractStakingIndexer == nil {
```
Comment (dustinxie) [inline chainservice/builder.go:336]: can simplify to:
```
if builder.cs.contractStakingIndexer == nil && len(SystemStakingContractAddress) > 0 {
    voteCalcConsts := builder.cfg.Genesis.VoteWeightCalConsts
    ....
}
```

File: chainservice/builder.go [hunk 2]
```diff
@@ -324,52 +324,50 @@ func (builder *Builder) buildContractStakingIndexer(forTest bool) error {
...
+			}
+			builder.cs.contractStakingIndexer = indexer
+		}
 	}
-	if forTest || builder.cfg.Genesis.SystemStakingContractV2Address == "" {
-		builder.cs.contractStakingIndexerV2 = nil
-		return nil
+	// build contract staking indexer v2
+	if builder.cs.contractStakingIndexerV2 == nil {
+		if builder.cfg.Genesis.SystemStakingContractV2Address == "" {
```
Comment (dustinxie) [inline chainservice/builder.go:359]: same here

File: db/kvstore_shards.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		if err := k.KVStore.Start(ctx); err != nil {
+			panic(fmt.Sprintf("failed to start underlying kvstore: %v", err))
+		}
+	})
+	return nil
+}
+
+func (k *kvStoreShards) Stop(ctx context.Context) error {
+	k.once.Do(func() {
+		if err := k.KVStore.Stop(ctx); err != nil {
```
Comment (dustinxie) [inline db/kvstore_shards.go:47]: `once` would only invoke `Do()` one time for the lifetime of the `once` instance, so the `Stop()` here won't be called

File: db/kvstore_shards.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	})
+	return nil
+}
+
+func (k *kvStoreShards) Shard(bucketPrefix string) KVStore {
+	return NewKVStoreWithPrefix(k, bucketPrefix)
+}
+
+// NewKVStoreWithPrefix creates a new KVStore with bucket prefix
+func NewKVStoreWithPrefix(kv KVStoreShards, bucketPrefix string) KVStore {
```
Comment (dustinxie) [inline db/kvstore_shards.go:59]: nit: can be `newKVStoreWithPrefix`

File: db/kvstore_shards.go (hunk 3 above)
Comment (dustinxie) [reply db/kvstore_shards.go:47]: need 2 `once`, one for Start and the other for Stop

File: db/kvstore_shards.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+func (k *kvStoreShards) Stop(ctx context.Context) error {
+	k.once.Do(func() {
+		if err := k.KVStore.Stop(ctx); err != nil {
+			panic(fmt.Sprintf("failed to stop underlying kvstore: %v", err))
+		}
+	})
+	return nil
+}
+
+func (k *kvStoreShards) Shard(bucketPrefix string) KVStore {
```
Comment (dustinxie) [inline db/kvstore_shards.go:54]: `Shard(name string)` is more clear?

File: db/kvstore_shards.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+	"github.com/pkg/errors"
+
+	"github.com/iotexproject/iotex-core/db/batch"
+)
+
+type (
+	// KVStoreShards is the interface of sharded KV store
+	KVStoreShards interface {
+		KVStore
+		Shard(bucketPrefix string) KVStore
```
Comment (dustinxie) [inline db/kvstore_shards.go:17]: `Shard(string)` is simpler

File: db/db_bolt.go [hunk 7]
```diff
@@ -56,10 +58,17 @@ func NewBoltDB(cfg Config) *BoltDB {
...
+	defer b.mutex.Unlock()
+
+	if b.IsReady() {
+		return nil
+	}
 	opts := *bolt.DefaultOptions
 	if b.config.ReadOnly {
 		opts.ReadOnly = true
 	}
+	var db *bolt.DB
```
Comment (CoderZhi) [inline db/db_bolt.go:71]: delete

File: systemcontractindex/stakingindex/index.go [hunk 8]
```diff
@@ -48,15 +48,21 @@ type (
...
 	}
 )
 
 // NewIndexer creates a new staking indexer
 func NewIndexer(kvstore db.KVStore, contractAddr string, startHeight uint64, blockInterval time.Duration) *Indexer {
+	bucketNS := stakingBucketNS + "#" + contractAddr
+	ns := stakingNS + "#" + contractAddr
 	return &Indexer{
-		common:        systemcontractindex.NewIndexerCommon(kvstore, stakingNS, stakingHeightKey, contractAddr, startHeight),
+		common:        systemcontractindex.NewIndexerCommon(kvstore, ns, stakingHeightKey, contractAddr, startHeight),
```
Comment (dustinxie) [inline systemcontractindex/stakingindex/index.go:61]: move these 2 `bucketNS, ns` into `newCache()`, so don't need to store them in `Indexer`, and L76 can keep the same

File: systemcontractindex/stakingindex/event_handler.go [hunk 9]
```diff
@@ -24,24 +24,28 @@ const (
...
+	tokenOwner      map[uint64]address.Address
+}
+
+func newEventHandler(ns, bucketNS string, dirty *cache) *eventHandler {
 	return &eventHandler{
-		dirty:      dirty,
-		delta:      batch.NewBatch(),
-		tokenOwner: make(map[uint64]address.Address),
+		stakingBucketNS: bucketNS,
+		stakingNS:       ns,
```
Comment (dustinxie) [inline systemcontractindex/stakingindex/event_handler.go:45]: add this new field, but i don't see where it's been used?

File: chainservice/builder.go (hunk 1 above)
Comment (dustinxie) [reply chainservice/builder.go:336]: fix this

**Review by dustinxie — APPROVED** [review]

**Pull request merged**